test/extensions/filters/http/compressor/compressor_filter_speed_test.cc

```diff
@@ -520,6 +567,120 @@ BENCHMARK(compressChunks1024WithZstd)
     ->UseManualTime()
     ->Unit(benchmark::kMillisecond);
 
+static std::vector<CompressionParams> brotli_compression_params = {
+    // level1 + default
+    {1, 0, 0, 0},
+
+    // level2 + default
+    {2, 0, 0, 0},
+
+    // level3 + default
+    {3, 0, 0, 0},
+
+    // level4 + default
+    {4, 0, 0, 0},
+
+    // level5 + default
+    {5, 0, 0, 0},
+
+    // level6 + default
+    {6, 0, 0, 0},
+
+    // level7 + default
+    {7, 0, 0, 0},
+
+    // level8 + default
+    {8, 0, 0, 0},
+
+    // level9 + default
+    {9, 0, 0, 0},
+
+    // level10 + default
+    {10, 0, 0, 0},
+
+    // level11 + default
+    {11, 0, 0, 0}};
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+static void compressFullWithBrotli(benchmark::State& state) {
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> decoder_callbacks;
```

Comment:
this is pre-existing but I've seen in the past that gmock structures are not built for speed, and may be a bottleneck in a benchmark.

Reply:
Yeah, it may be a issue, but I don't have any other idea to address it for now. And we more care about relative values of different compression algorithms in our work.

Reply:
you can simply use the real objects. See other *_speed_test.cc around the hierarchy for examples. That's the preferred solution.

But at a minimum you could get some performance data (eg flame graphs) to see whether you are really just measuring the speed of the slow mocks. When you do this, please use both optimization *and* debug so you get accurate symbols in your flame graphs.

I don't think you can learn enough about relative speed values if the time is dominated by mocks. I'm not sure it is, but I can't really trust benchmark data unless we get some better view.

Reply:
![2022-10-13 15-35-56 的屏幕截图](https://user-images.githubusercontent.com/20723282/195533290-b8d4ad5f-803a-4411-ad9d-ece637f75a3e.png)
I have got the result of flamegraph, it seems `encodeData` takes up more than 80% of the time, and `Mock` is about 7% totally.

Reply:
Did you take that flame-graph from an optimized build? Actually I'd recommend a build with debug *and* optimization: `--compilation_mode=opt --cxxopt=-g --cxxopt=-ggdb3`.

But let's take that 7% overhead and do some math. Some say (and I think that used to be true) that Brotli is 10x slower than gzip.

So if Brotli takes 93 seconds to do an operation, with 7 seconds for mocks, and Gzip takes 9.3 seconds, with 7 seconds per mocks, your perf data will look like the delta is 100/(9.3+7) = 6x slower when it's really 10x slower.

That's why I think it's worth getting rid of the mock overhead by using real objects.

I also think it's OK for benchmarks to skip the filtering infrastructure and just call the common compressor API, which might reduce the number of objects you need to instantiate.

Reply:
I rebuild with `--compilation_mode=opt --cxxopt=-g --cxxopt=-ggdb3`, it looks that the percentage of `encodeData` in case of gzip and brotli is increased up to 96%, and `Mock` is down to about 1.5%.
![2022-10-14 10-35-24 的屏幕截图](https://user-images.githubusercontent.com/20723282/195767868-2f7181ac-cc20-432d-be28-cc24f0df1300.png)
![2022-10-14 10-38-21的屏幕截图](https://user-images.githubusercontent.com/20723282/195767890-d215876c-a525-414c-99ef-6f67005ea21f.png)

And the percentage of `encodeData` in case of zstd is increased up to 92%, and `Mock` is down to about 3%.
![2022-10-14 10-36-15 的屏幕截图](https://user-images.githubusercontent.com/20723282/195767897-db3cf2af-8695-4618-8a0f-241f88b3e6b6.png)

I think that the time of `Mock` is related to the total time of benchamrk, and its percentage is roughly fixed in one benchmark.

Reply:
OK I think this is acceptable but please comment that the gmock overhead has been measured with flame graphs to be pretty low.

I think the reason it was 7% earlier was that optimization affects different code-paths differently, and in particular heavy use of STL really needs the optimizer to work fast, whereas more C-like code (e.g. found in zlib or brotli) may is less reliant on optimization. When you compiled with some optimization you reduced the relative overhead of STL-heavy mocks. That's my theory anyway.

Reply:
OK, I have added comments. 
I learned a lot from your comments. 
Thank you very much! :)